### DISCUSSION TITLE
Let us programmatically set the current Tab (or better yet expose the context)

### DISCUSSION THREAD

**Discussion by mattiaz9** [tailwindlabs/headlessui#742] [Ideas]
Right now switching between tabs is only possibile with mouse and keyboard, but there are case where it would be useful to manually set the current tab in the code (for instance, a 'Next' button that should select the following tab).

I'm not sure what could be the best way to achieve it, but for what I could've seen in the code the components use the react context, and maybe the package could expose a `Tabs.TabsContext` that we can use for our purposes.

**Comment by diegosanw** [comment]
This would be great!

My use case is this one:
I want to change the TabList for a dropdown menu on small screens. I would like to keep using the same TabPanels and switch the current TabPanel with the dropdown change event.

The only solution I found so far is to replicate the contents of the TabPanels and not use headlessui for small screens.

**Reply by yannsheriff** [reply]
It would be awesome, I have the same use case here.

**Reply by dospunk** [reply]
What we've done for this is to hide the <TabList> on small screens but not the <TabPanels>, and have a <select> that programmatically clicks the corresponding <Tab> when its value is changed. It feels a little hacky to write but it works like a charm

**Comment by dospunk** [comment]
It is technically possible, but kind of hacky. In Vue terms, you can do it by keeping an array of `ref`s to each of the Tab buttons, then you can programmatically click on the tabs through their `ref`s.

**Comment by LoganDark** [comment]
It surprises me that the Tabs component is not controlled. I wish they had chosen to use `index` rather than `defaultIndex`

**Comment by devinfd** [comment]
I have a current need for this feature for the same reason @ds-perfit describes

**Comment by hjbdev** [comment]
I don't really see much reason why the context should not be exposed for all components. This library is one of the best UI component libraries out there, but the lack of programmatic control can be frustrating at times.

**Comment by jamesray** [comment]
Yeah the ability to manipulate the context without simulating a click would be great for so many use-cases.

**Reply by Regisnut** [reply]
same issue, I 'd like to put link in my footer and onclick to get to the tabs required and I'm stuck with only defaultIndex !

**Comment by Regisnut** [comment]
pity, with chakra ui, you have something like this :
 <Tabs index={tabIndex} 
 ...

**Comment by JacobReynolds** [comment]
I'd be curious to know if there are any semantic or accessibility reasons for not having this. If not, I'd be happy to try and open a pull request with the functionality. I've got a good example of it working for react here https://github.com/NetSPI/headlessui, but it could probably be more robust.

**Comment by OmkarK45** [comment]
One good use case is creating step-forms ! One can have next/previous buttons to programmatically change the current tab

**Reply by seanparkonline** [reply]
In combo with Formik it's magic for multi-step forms!
🪄

**Reply by husleuujii** [reply]
is there any repo?

**Comment by thiagonzalez** [comment]
I'm in the same situation here. I have links in tab 1, which needs to change the tab on click to tab 2 or 3:

![image](https://user-images.githubusercontent.com/967571/139471343-533cc2c4-6d1d-432d-8e07-6a6afbcbb930.png)

I can see the state changing, but the `defaultIndex` doesn't reflect my state. I'm out of ideas here :(
The only way to fix it here is to refresh the page after adding a queryString to the URL.

**Reply by thiagonzalez** [reply]
guys, @marcioj found a solution! It's not exactly the best way to do it but solved the issue.

I have a const `currentTabIndex` with the value of the current tab I want to display.
I'm observing the queryString changing and updating the `currentTabIndex` to the new value.

If I just use `defaultIndex` on Tabs.Group, the component won't change, but adding a `key` prop to it does the magic.

```
<TabGroup
  key={currentTabIndex}
  defaultIndex={currentTabIndex}
>
```

Again, that's not the best way to fix the problem, because the whole component will re-render.

**Reply by Regisnut** [reply]
many thanks, this has fixe my problem ! react key works for me in Tabgroup.

**Reply by yordis** [reply]
Same deal for us, https://github.com/tailwindlabs/headlessui/discussions/902

We had to use the `key` just so we trigger a re-rendering 😭

**Reply by seanparkonline** [reply]
Thanks for this. Works. 
Hope Tailwind add support for the feature soon.

**Reply by dwome** [reply]
Great and clean way!

**Reply by Desalubrious** [reply]
Thank you! This saved me hours of headache.

**Reply by DanielBailey-web** [reply]
@Desalubrious there is a built in way:
https://github.com/tailwindlabs/headlessui/discussions/742#discussioncomment-1998282

The way mentioned above completely re-renders the component due to the change in the key which generally is not what you are looking for.

**Comment by seanparkonline** [comment]
As per @thiagonzalez and @marcioj, tracking the `selectedTabIndex` in state and then passing it to `<Tab.Group>` as a `key` prop to trigger a re-render works. 

One extra note is that you should also use the `onChange` prop to set the index of your state so that if the user navigates by keyboard or clicking on the tab the state is updated also. 
eg:

```jsx
<Tab.Group key={selectedTabIndex} defaultIndex={selectedTabIndex} onChange={index => setSelectedTabIndex(index)}>
    ...
```
Hope Tailwind adds native support for this soon.

**Comment by DanielBailey-web** [comment]
This was added, you can do this:
```js
<Tab.Group
  onChange={(i: any) => setTabIndex(i)}
  selectedIndex={tabIndex}
>
  {children}
</Tab.Group>
```

**Reply by rjlg** [reply]
Hey @DanielBailey-web where is setTabIndex imported from? 

https://github.com/tailwindlabs/headlessui/search?q=setTabIndex the only reference to setTabIndex I can find in this repo is your comment.

**Reply by DanielBailey-web** [reply]
@rjlg its from a useState hook. Sorry I didn't make that clear. 
`const [tabIndex, setTabIndex] = useState(0)`

You could of course use any type of state and way to set it (such as class component state, or context [useContext, redux or other] to accomplish this)

**Reply by dospunk** [reply]
On Vue `selectedIndex` is a slot prop, so this method will only work on React it seems

**Reply by DanielBailey-web** [reply]
@dospunk Look at my edit for react, there is a change Vue was already using tabIndex as the prop. (Idk tho)

**Reply by RobinMalfait** [reply]
@DanielBailey-web this is incorrect, we did have a `defaultIndex` prop, this has not been changed it still exists. The prop to control data is `selectedIndex` for both `React` and `Vue`. `tabIndex` _also_ exists, but this controls the native `tabindex` on focusable elements.

See: https://github.com/tailwindlabs/headlessui/discussions/742#discussioncomment-2335074 for more info.

**Reply by DanielBailey-web** [reply]
@RobinMalfait  Oh shoot you right my B. Edited it back. But I get a type error when using selectedIndex now... (but it still works)

**Comment by RobinMalfait** [comment]
Hey all!

The Tabs component can be a controlled component now, this means that this is possible out of the box using the `selectedIndex` prop.

More info can be found here:
- React: https://headlessui.dev/react/tabs#controlling-the-active-tab
- Vue: https://headlessui.dev/vue/tabs#controlling-the-active-tab

**Reply by leon-up9** [reply]
how to reset the selected tab to be nothing?(select none of the tabs)

**Comment by maupeon** [comment]
<img width="912" alt="Captura de Pantalla 2022-05-02 a la(s) 1 44 36 p m" src="https://user-images.githubusercontent.com/17842403/166306370-2dbe679f-61f2-4d4e-9144-74e7cb5a6441.png">

This is happening to me when trying to control selectedIndex

**Reply by DanielBailey-web** [reply]
Post your code :eyes:

**Comment by leon-up9** [comment]
how to reset the selected tab to be nothing?(select none of the tabs)

**Reply by pranto-rayple** [reply]
Just pass nothing in the useState()
